### PR TITLE
[stable] Backport fix for #60989 "ICE with incorrect turbofish"

### PR DESCRIPTION
**Pull request by oli-obk**
stable fix backport for #60989 

do not merge yet, untested

### CODE REVIEW

**Comment by eddyb** [comment]
Started stage1 test build to help out, apparently LLVM builds make it hard to test stable backports.

File: src/test/run-pass/issue-60989.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+fn main() {
+    let c1 = ();
+    c1::<()>;
```
Comment (eddyb) [inline src/test/run-pass/issue-60989.rs]: Wait how would this compile successfully?!

Comment (oli-obk) [reply src/test/run-pass/issue-60989.rs]: oops
it would not
should have been a compile-fail test

**Comment by eddyb** [comment]
```
error: internal compiler error: src/librustc/hir/def.rs:271: attempted .def_id() on invalid def: Local(NodeId(13246))
```
I talked to @oli-obk, and he doesn't mind me taking over (since I already have a build dir for this).

**Comment by pietroalbini** [comment]
This is ok from a release point of view, but I'm not knowledgeable enough to be the assignee for this.

File: src/test/compile-fail/issue-60989.rs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+fn main() {
+    let c1 = ();
+    c1::<()>;
```
Comment (Centril) [inline src/test/compile-fail/issue-60989.rs]: We might also want to extend this with the other forms in the issue?

File: src/test/compile-fail/issue-60989.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+fn main() {
```
Comment (Centril) [inline src/test/compile-fail/issue-60989.rs]: Also -- don't we need an `.stderr` file?

File: src/test/compile-fail/issue-60989.rs (hunk 2 above)
Comment (oli-obk) [reply src/test/compile-fail/issue-60989.rs]: will do on master. This is just for stable.

File: src/test/compile-fail/issue-60989.rs (hunk 3 above)
Comment (oli-obk) [reply src/test/compile-fail/issue-60989.rs]: only ui tests need those

Comment (Centril) [reply src/test/compile-fail/issue-60989.rs]: oh

File: src/test/compile-fail/issue-60989.rs (hunk 2 above)
Comment (Centril) [reply src/test/compile-fail/issue-60989.rs]: Yeah good idea 👍 -- Let's leave this here as a reminder.

**Comment by eddyb** [comment]
`./x.py test --stage 1 src/test/ui` succeeds now for me, what's left?
Presumably someone other than me has to review it (@varkor?).

**Comment by pietroalbini** [comment]
Anyone from @rust-lang/compiler up for a review of this?

**Comment by cuviper** [comment]
There's no such label, but let's not forget this for master/nightly too...

**Comment by varkor** [comment]
@cuviper: https://github.com/rust-lang/rust/issues/60989 isn't being marked as fix, so we have the issue to remind us 👍

**Comment by cuviper** [comment]
I don't know if this is spurious, but the rls test failure is here:
https://ci.appveyor.com/project/rust-lang/rust/builds/24765767/job/c72eopevht8wxajk#L10885

    thread 'client_changing_workspace_lib_retains_diagnostics' panicked at 'assertion failed: bin.diagnostics[0].message.contains("mismatched types\n\nexpected u32, found u64")', src\tools\rls\tests\client.rs:275:5

**Comment by varkor** [comment]
(I'm also not sure, but this is blocking the stable release, so I'm going to hope it was.)

**Comment by eddyb** [comment]
cc @Xanewok for https://github.com/rust-lang/rust/pull/61085#issuecomment-495351805

**Comment by pietroalbini** [comment]
Yeah that should be spurious.

**Comment by pietroalbini** [comment]
Unfortunately we can't include this into 1.35.0: the spurious failure moved the tentative release time to around 2AM in my timezone, and y'all don't want me doing a release that late. @eddyb or @oli-obk, can any of you rebase this PR on master and beta-nominated+stable-nominated it?

Closing this PR.

@bors r- retry

**Comment by eddyb** [comment]
Note for whoever wants to rebase this on master (before I have a chance to look at it): `Def` was renamed to `Res`, so at the very least there's that, but also `rewrite_self_ctor` might have changed its signature and may be more aggressive now. (if so, it should be changed to return just `Res`)

**Pull request closed without merging**